src/library_wasmfs.js

```diff
@@ -85,6 +86,37 @@ var WasmfsLibrary = {
       // For file preloading, cwd should be '/' to begin with.
       return '/';
     }
+  },
+  _emscripten_write_js_file: function(index, buffer, length, offset) {
+    var size = wasmFS$JSMemoryFiles[index].length;
+    // Resize the typed array if the length of the write buffer exceeds its capacity.
+    if (offset + length >= size) {
+      var oldContents =wasmFS$JSMemoryFiles[index];
+      wasmFS$JSMemoryFiles[index] = new Uint8Array(offset + length);
+      wasmFS$JSMemoryFiles[index].set(oldContents.subarray(0, size));
+    }
+
+    wasmFS$JSMemoryFiles[index].set(HEAPU8.subarray(buffer, buffer + length), offset);
+  },
+  _emscripten_read_js_file: function(index, buffer, length, offset) {
+    HEAPU8.set(wasmFS$JSMemoryFiles[index].subarray(offset, offset + length), buffer);
+  },
+  _emscripten_get_js_file_size: function(index) {
+    return wasmFS$JSMemoryFiles[index].length;
+  },
+  _emscripten_create_js_file: function() {
+    // Traverse wasmFS$JSMemoryFiles until an empty index is found.
+    for (var index = 0; index < wasmFS$JSMemoryFiles.length; index++) {
+      if (wasmFS$JSMemoryFiles[index] === null) {
+        wasmFS$JSMemoryFiles[index] = new Uint8Array(0);
+        return index;
+      }
+    }
+    wasmFS$JSMemoryFiles.push(new Uint8Array(0));
```

Comment:
This line can be removed - we could allocate on the first write. Then we'd know a useful size to allocate on.

Reply:
Is that worth adding an edge case for reading from zero-sized files? (Or would the code handle that gracefully as-is? I'm not sure)

Reply:
We would need to add additional error checking for the zero-sized file case. I don't think it would be an improvement and it's more intuitive to just create an empty sized array and access its length.

Reply:
Fair points. However, if it's an extra line of code either way, I think we should learn towards the more efficient one. Avoiding allocating a typed array is something that we try to do as much as possible in general.

I am hoping that the extra code the other way would be shorter, too - basically `if (!buffer || buffer is too small) allocate buffer`